File: Day-11/LSRK.py
from collections import deque

def is_repeated_subsequence(s: str, seq: str, k: int) -> bool:
    pos = 0
    for _ in range(k):
        for ch in seq:
            pos = s.find(ch, pos)
            if pos == -1:
                return False
            pos += 1
    return True

def longest_subsequence_repeated_k(s: str, k: int) -> str:
    # Count character frequency
    freq = [0] * 26
    for ch in s:
        freq[ord(ch) - ord('a')] += 1

    # Characters that appear at least k times
    available_chars = [chr(i + ord('a')) for i in range(26) if freq[i] >= k]

    result = ""
    queue = deque([""])

    while queue:
        curr = queue.popleft()

        for ch in available_chars:
            new_seq = curr + ch

            if is_repeated_subsequence(s, new_seq, k):
                if len(new_seq) > len(result) or (len(new_seq) == len(result) and new_seq > result):
                    result = new_seq
                queue.append(new_seq)

    return result

# Example usage
print(longest_subsequence_repeated_k("letslets", 2))  # Output: "lets"

File: Day-11/LSRK.go
package main

import (
	"fmt"
	"strings"
)

func isRepeatedSubsequence(s string, seq string, k int) bool {
	pos := 0
	for i := 0; i < k; i++ {
		for _, c := range seq {
			idx := strings.IndexByte(s[pos:], byte(c))
			if idx == -1 {
				return false
			}
			pos += idx + 1
		}
	}
	return true
}

func longestSubsequenceRepeatedK(s string, k int) string {
	freq := make([]int, 26)
	for _, c := range s {
		freq[c-'a']++
	}

	var availableChars []rune
	for c := 'a'; c <= 'z'; c++ {
		if freq[c-'a'] >= k {
			availableChars = append(availableChars, c)
		}
	}

	result := ""
	queue := []string{""}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]

		for _, c := range availableChars {
			next := curr + string(c)

			if isRepeatedSubsequence(s, next, k) {
				if len(next) > len(result) || (len(next) == len(result) && next > result) {
					result = next
				}
				queue = append(queue, next)
			}
		}
	}

	return result
}

func main() {
	fmt.Println(longestSubsequenceRepeatedK("letslets", 2)) // Output: "lets"
}